Program.cs

```diff
@@ -1,41 +1,53 @@
 ﻿static void TreasureHunt()
 {
-    var (gridRows, gridColumns) = GetGridSize();
-    var grid = InitializeGrid(gridRows, gridColumns);
-    var numberOfTreasures = PlaceTreasures(grid);
-    var numberOfObstacles = PlaceObstacles(grid);
-    var (playerRow, playerColumn) = PlacePlayer(grid);
-    var (monsterRow, monsterColumn) = PlaceMonster(grid);
-
-    var treasureCounter = 0;
-    Console.Clear();
-    DisplayGrid(grid);
-
-    ConsoleKeyInfo keyInfo;
-    while ((keyInfo = Console.ReadKey(true)).Key != ConsoleKey.Escape)
-    {
-        (playerRow, playerColumn, treasureCounter) =
-            MovePlayer(grid, playerRow, playerColumn, keyInfo, treasureCounter);
-        (monsterRow, monsterColumn) = MoveMonster(grid, playerRow, playerColumn, monsterRow, monsterColumn);
+    var gameRunning = true;
 
+    while (gameRunning)
+    {
+        var (gridRows, gridColumns) = GetGridSize();
+        var grid = InitializeGrid(gridRows, gridColumns);
+        var numberOfTreasures = (int)(gridRows * gridColumns * 0.1);
+        var numberOfObstacles = (int)(gridRows * gridColumns * 0.2);
+        var (playerRow, playerColumn) = PlacePlayer(grid);
+        var (monsterRow, monsterColumn) = PlaceMonster(grid);
+
+        var treasureCounter = 0;
         Console.Clear();
         DisplayGrid(grid);
-        Console.WriteLine("Number of Treasures: " + numberOfTreasures);
-        Console.WriteLine("Treasure Counter: " + treasureCounter);
+        PlaceItems(numberOfTreasures, '$', grid);
+        PlaceItems(numberOfObstacles, '|', grid);
 
-        if (treasureCounter == numberOfTreasures)
-        {
-            Console.Clear();
-            Console.WriteLine($"You found all {numberOfTreasures} treasures and win the game!");
-            break;
-        }
+        ConsoleKeyInfo keyInfo;
+        var gameActive = true;
 
-        if (monsterRow == playerRow && monsterColumn == playerColumn)
+        while (gameActive && (keyInfo = Console.ReadKey(true)).Key != ConsoleKey.Escape)
```

Comment:
We usually try to avoid grouping assignments and conditions as it decreases the readability.